samples/bluetooth/alexa_gadget/src/gadgets/CMakeLists.txt

```diff
@@ -0,0 +1,187 @@
+    message(STATUS "Please download ${NANOPB_REPO_TAG} from here https://jpa.kapsi.fi/nanopb/download/")
+    message(STATUS "and extract to either this sample directory or somewhere cmake can find it via system PATH.")
+    message(FATAL_ERROR "${NANOPB_REPO_TAG} required by this project.")
+endif()
+
+message(STATUS "Using protoc: ${PROTOC_BIN}")
+message(STATUS "Using protoc-gen-nanopb: ${NANOPB_BIN}")
+message(STATUS "Using nanopb source tag: ${NANOPB_REPO_TAG}")
+
+function(compile_protobufs_in_dir dir recursive)
+    # Find all .proto files and include directories
+    if (recursive)
+        file(GLOB_RECURSE PROTO_FILES "${dir}/*.proto")
+    else()
+        file(GLOB PROTO_FILES "${dir}/*.proto")
+    endif()
+
+    foreach(file ${PROTO_FILES})
+        get_filename_component(FILE_PATH ${file} DIRECTORY)
+        list(APPEND PROTOC_INC_DIRS "-I${FILE_PATH}")
+    endforeach()
+
+    # Compile .proto files in their respective directories
+    foreach(file ${PROTO_FILES})
+        # Add all directories containing .proto files.
+        # Duplicates and current directory is then removed from "-I" arguments
+        set(PROTOC_ARGS "${PROTOC_INC_DIRS}")
+        get_filename_component(FILE_PATH ${file} DIRECTORY)
+        get_filename_component(FILE_NAME ${file} NAME)
+        list(REMOVE_ITEM PROTOC_ARGS "-I${FILE_PATH}")
+        list(REMOVE_DUPLICATES PROTOC_ARGS)
+        list(APPEND PROTOC_ARGS "-I.")
+        list(APPEND PROTOC_ARGS "--nanopb_out=${FILE_PATH}")
+        list(APPEND PROTOC_ARGS "${FILE_NAME}")
+        message(STATUS "Compiling proto definition: ${FILE_NAME}")
+        execute_process(
+            COMMAND ${PROTOC_BIN} ${PROTOC_ARGS}
+            WORKING_DIRECTORY ${FILE_PATH}
+        )
+    endforeach()
+endfunction()
+
+include(FetchContent)
+
+# Get repo containing protobuf definitions
+FetchContent_Declare(
+    gadgets_repo
+    GIT_REPOSITORY ${GADGETS_PROTO_REPO}
+    GIT_TAG        ${GADGETS_PROTO_TAG}
+)
+
+FetchContent_GetProperties(gadgets_repo)
+if(NOT gadgets_repo_POPULATED)
+    FetchContent_Populate(gadgets_repo)
+endif()
+
+if (NOT PROTOBUF_GENERATION_PERFORMED)
+    # Only compile protobuf definitions at first build to reduce rebuild time.
+    # The only protobuf definition that can be changed is the "custom event" type,
+    # and this is not expected to be updated often.
+
+    # Compile all .proto files found in gadgets_repo
+    compile_protobufs_in_dir(${gadgets_repo_SOURCE_DIR} TRUE)
+
+    # Compile all .proto files found in sample "proto" directory, if any
+    compile_protobufs_in_dir(${CMAKE_CURRENT_SOURCE_DIR}/proto FALSE)
+    set(PROTOBUF_GENERATION_PERFORMED TRUE CACHE INTERNAL "")
+endif()
+
+# Get repo containing nanopb source files
+FetchContent_Declare(
+    nanopb_repo
+    GIT_REPOSITORY ${NANOPB_REPO}
+    GIT_TAG        ${NANOPB_REPO_TAG}
+)
+
+FetchContent_GetProperties(nanopb_repo)
+if(NOT nanopb_repo_POPULATED)
+    FetchContent_Populate(nanopb_repo)
+endif()
```

Comment:
we should not dowload code as part of CMake (or build).

This should go into the manifest file instead.

Reply:
nanopb might be useful for other samples/applications/libraries in the future, and makes sense to include in the west manifest. However, the alexa .proto-files are very sample specific. The reason for not including these files in the sample directly is to avoid the hassle of vetting software license terms for redistribution.

Reply:
> The reason for not including these files in the sample directly is to avoid the hassle of vetting software license terms for redistribution.

I'm not talking about commiting those files into our repo.
Only that we do not fetch them as part of CMake, but instead as part of `west`.

Regarding licensing terms, I don't see the difference in `west` downloading the files, compared to `CMake` downloading the files. Can you explain ?

Reply:
Maybe I'm misunderstanding. Are you suggesting adding the two repos to west.yml in the top-level nrf directory, or creating a manifest file under this sample?

I agree that there is no difference in having west download the files compared to having cmake download them. I thought the benefit of using the CMake FetchContent functionality would be that the files are only downloaded when one wants to build this particular sample, and not every time one runs west update.

Reply:
> Are you suggesting adding the two repos to west.yml in the top-level nrf directory

Yes, that way we keep a single overview of all repos we use.
Secondly, it has a benefit for users to do: `west update` and then take their laptop with them on travel without requiring network access (as example on a plane).
Couple of more reasons.
- By starting to use CMake FetchContent, their is no longer a common place to look for repos.
- You start depending on network access at build time.
- This would be the first place to use CMake FetchContent in Zephyr and NCS.

> I thought the benefit of using the CMake FetchContent functionality would be that the files are only downloaded when one wants to build this particular sample, and not every time one runs west update.

We still download `openthread` even if user is only interessted in LTE-M/NB-IOT. 
Tinycrypt, even if users only needs cc310,  and so on.

If we should start to download code as part of CMake build, then that should be taken as a wider discussion.

Reply:
I will add these two repos to the west manifest as suggested